Actions: API, command, and jobspec docs (#19166)
hashicorp/nomad: website/content/docs/job-specification/action.mdx

---
layout: docs
page_title: action Block - Job Specification

> description: The "action" block allows for executable commands that operators

Reviewer: The sentence seems to be missing the end.

---

# `action` Block

<Placement groups={['job', 'group', 'task', 'action']} />

The `action` block allows job authors to define custom commands. These commands
can be executed by operators with the necessary permissions on a running
allocation, offering a controlled way to interact with tasks.

## `action` Parameters

> The `action` block is used within a task and has the following parameters:

Reviewer: ```suggestion
```

This is shown in the `Placement` header.

> - `command` (Required): Specifies the command to be executed.
> - `args` (Optional): Provides a list of arguments to pass to the command.

Reviewer: ```suggestion
- `command` `(string: <required>)` - Specifies the command to be executed.
- `args` `(array<string>: [])`: Provides a list of arguments to pass to the command.
```

We have a weird mishmash of how these are defined, but this is the pattern we use for job spec.

```hcl
job "my-job" {
  group "my-group" {
    task "my-task" {
      action "get-changelog" {
        command = "/usr/bin/curl"

>         args = ["-s", "https://raw.githubusercontent.com/hashicorp/nomad/main/CHANGELOG.md"]

Reviewer: ```suggestion
        args = [
          "-s", 
          "https://raw.githubusercontent.com/hashicorp/nomad/main/CHANGELOG.md",
        ]
```

Just to reduce line length and avoid horizontal scrolling

>       }

Reviewer: ```suggestion
      }
      # ...
```

Indicate that there are more stuff expected inside the task.

    }
  }
}
```

## `action` Examples

### Basic Action
This example demonstrates a simple action that prints the current date and time:

```hcl

> job "example" {
>   ...
>   group "demo" {
>     ...
>     task "show-date" {
>       ...
>       action "current-date" {
>         command = "/bin/date"
>       }
>     }
>   }
> }

Reviewer: ```suggestion
job "example" {
  # ...
  group "demo" {
    # ...
    task "show-date" {
      # ...
      action "current-date" {
        command = "/bin/date"
      }
    }
  }
}
```

Take advantage of the syntax highlighter to render these nicely.

```

### Action with Arguments
Here, an action uses arguments to perform a specific task:

```hcl

> job "example" {
>   ...
>   group "demo" {
>     ...
>     task "list-files" {
>       ...
>       action "list-tmp" {
>         command = "/bin/ls"
>         args    = ["-l", "/tmp"]
>       }
>     }
>   }
> }

Reviewer: ```suggestion
job "example" {
  # ...
  group "demo" {
    # ...
    task "list-files" {
      # ...
      action "list-tmp" {
        command = "/bin/ls"
        args    = ["-l", "/tmp"]
      }
    }
  }
}
```

Same as above but GitHub was not happy to have codeblocks inside the suggestion 😅

```

### Action with Template
This advanced example demonstrates an action that fetches and formats the latest
changelog from the Nomad GitHub repository using a shell script. It showcases
the use of templating with embedded environment variables and a multi-line script.

```hcl
action "fetch-latest-nomad-changelog" {
  command = "/bin/sh"
  args    = ["-c", 

>     <<EOT

Reviewer: TIL about using heredocs as a list element, never thought about this before 😄
